### ISSUE TITLE
Verified le-wm inference & training on AMD MI300X (ROCm 6.4) — steps & results

### ISSUE THREAD

**Issue by ZJLi2013** [lucas-maes/le-wm#15]
Hi, I successfully ran both pretrained inference and from-scratch training of LeWM on AMD Instinct MI300X GPUs with ROCm. Sharing the setup steps and results here in case it helps others looking to reproduce on AMD hardware.

## 1. Using HuggingFace checkpoints with eval.py

The HuggingFace model repos (`quentinll/lewm-*`) store weights-only checkpoints (`weights.pt` + `config.json`), while `eval.py` expects serialized JEPA objects (`*_object.ckpt`). I wrote a small conversion script to bridge the two:

<details>
<summary><b>convert_hf.py</b> (click to expand)</summary>

```python
"""Download LeWM checkpoints from Hugging Face and convert to eval-compatible format."""

import argparse, json, os
import torch
import stable_pretraining as spt
from huggingface_hub import hf_hub_download
from jepa import JEPA
from module import ARPredictor, Embedder, MLP

HF_REPOS = {
    "pusht":    {"repo": "quentinll/lewm-pusht",    "output_dir": "pusht"},
    "tworooms": {"repo": "quentinll/lewm-tworooms", "output_dir": "two-room"},
    "cube":     {"repo": "quentinll/lewm-cube",     "output_dir": "cube"},
    "reacher":  {"repo": "quentinll/lewm-reacher",  "output_dir": "reacher"},
}

def build_model_from_config(cfg):
    enc = cfg["encoder"]
    encoder = spt.backbone.utils.vit_hf(
        enc["scale"], patch_size=enc["patch_size"],
        image_size=enc["image_size"], pretrained=False, use_mask_token=False,
    )
    return JEPA(
        encoder=encoder,
        predictor=ARPredictor(**cfg["predictor"]),
        action_encoder=Embedder(**cfg["action_encoder"]),
        projector=MLP(**cfg["projector"], norm_fn=torch.nn.BatchNorm1d),
        pred_proj=MLP(**cfg["pred_proj"], norm_fn=torch.nn.BatchNorm1d),
    )

def convert_task(task, cache_dir):
    info = HF_REPOS[task]
    config_path = hf_hub_download(repo_id=info["repo"], filename="config.json")
    weights_path = hf_hub_download(repo_id=info["repo"], filename="weights.pt")
    with open(config_path) as f:
        cfg = json.load(f)
    model = build_model_from_config(cfg)
    model.load_state_dict(torch.load(weights_path, map_location="cpu", weights_only=True))
    model.eval()
    out_dir = os.path.join(cache_dir, info["output_dir"])
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "lewm_object.ckpt")
    torch.save(model, out_path)
    print(f"Saved: {out_path}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--tasks", nargs="+", choices=list(HF_REPOS), default=list(HF_REPOS))
    parser.add_argument("--cache-dir", default=os.environ.get("STABLEWM_HOME", os.path.expanduser("~/.stable-wm")))
    args = parser.parse_args()
    for task in args.tasks:
        convert_task(task, args.cache_dir)
```

</details>

Usage:

```bash
pip install huggingface-hub
python convert_hf.py                        # all 4 tasks
python convert_hf.py --tasks pusht tworooms # specific tasks
```

This produces the `_object.ckpt` files that `eval.py` / `AutoCostModel` expect.

## 2. AMD MI300X environment setup

**Hardware:** 8× AMD Instinct MI300X  
**Software:** ROCm 6.3, PyTorch 2.9.1+rocm6.3, Python 3.12

### Docker container

```bash
docker run -d --name lewm_rocm \
  --device=/dev/kfd --device=/dev/dri \
  --group-add video --ipc=host --shm-size 16G \
  --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
  -v $(pwd)/le-wm:/workspace/le-wm \
  -w /workspace/le-wm \
  rocm/pytorch:rocm6.4_ubuntu24.04_py3.12_pytorch_release_2.6.0 \
  sleep infinity
```

### Dependency installation

```bash
# System deps
apt-get update -qq && apt-get install -y -qq swig cmake zstd

# Python deps
pip install uv && uv pip install --system stable-worldmodel[train,env] huggingface-hub datasets

# IMPORTANT: reinstall ROCm PyTorch (pip install above overrides it with the CUDA version)
pip uninstall -y torch torchvision triton
pip install torch torchvision --index-url https://download.pytorch.org/whl/rocm6.3
```

> **Note:** `stable-worldmodel` pulls in a CUDA-only `torch` from PyPI, silently replacing the ROCm build. The reinstall step above is required on AMD GPUs.

### Data download (from HuggingFace)

```bash
# Checkpoints → _object.ckpt
python convert_hf.py

# PushT dataset (~13 GB compressed → 46 GB)
python -c "from huggingface_hub import hf_hub_download; hf_hub_download('quentinll/lewm-pusht', 'pusht_expert_train.h5.zst', repo_type='dataset', local_dir='/tmp/hf')"
zstd -d /tmp/hf/pusht_expert_train.h5.zst -o ~/.stable-wm/pusht_expert_train.h5

# TwoRoom dataset (~3.4 GB compressed → 12.8 GB)
python -c "from huggingface_hub import hf_hub_download; hf_hub_download('quentinll/lewm-tworooms', 'tworoom.tar.zst', repo_type='dataset', local_dir='/tmp/hf')"
tar --zstd -xvf /tmp/hf/tworoom.tar.zst -C ~/.stable-wm/
```

## 3. Inference results

No code changes were needed — le-wm runs out of the box on ROCm.

```bash
MUJOCO_GL=egl python eval.py --config-name=tworoom.yaml policy=two-room/lewm eval.num_eval=3
MUJOCO_GL=egl python eval.py --config-name=pusht.yaml  policy=pusht/lewm   eval.num_eval=3
```

| Task | eval.num_eval | success_rate | CEM solve time |
|------|:---:|:---:|---|
| TwoRoom | 3 | **100%** | 15.2s (1st), ~10s (subsequent) |
| PushT | 3 | **100%** | 13.0s (1st), ~10s (subsequent) |

## 4. Training results

Training works with 8-GPU DDP out of the box. One minor compatibility fix needed:

> `stable_pretraining==0.1.4` + `lightning==2.6.1`: `self.optimizers()` returns a single optimizer instead of a list, causing `len()` to fail in `on_train_start`. One-line fix:
> ```python
> # in stable_pretraining/module.py, line 253
> optimizers = self.optimizers()
> optimizers = optimizers if isinstance(optimizers, list) else [optimizers]  # ← add this
> ```

```bash
python train.py data=pusht wandb.enabled=False trainer.max_epochs=2 loader.batch_size=32
```

| Metric | Value |
|---|---|
| final loss | 0.134 |
| pred_loss | 0.019 |
| sigreg_loss | 1.278 |
| val/pred_loss | 0.014 |
| throughput | 10.56 it/s |
| time per epoch | ~10 min 49 sec |
| GPU utilization | 95–99% (all 8 GPUs) |
| VRAM usage | ~4% per GPU |
| checkpoints saved | `lewm_epoch_{1,2}_object.ckpt`, `lewm_weights.ckpt` |

## Summary

- **Inference:** works on AMD MI300X + ROCm 6.3 with zero code changes, HF checkpoints need the conversion script above.
- **Training:** works with 8-GPU DDP, bf16 precision. Only blocker was a minor `stable_pretraining` / Lightning 2.6 compat issue (not in this repo).
- **Key gotcha:** `pip install stable-worldmodel` replaces ROCm PyTorch with CUDA PyTorch — needs manual reinstall.

Great work on making the codebase device-agnostic! 🎉

**Comment by quentinll** [comment]
Hey @ZJLi2013, thank you for this nice experiment! I will update the installation guidelines for AMD hardware

**Comment by Edingburge** [comment]
分享一个楼主的离线加载脚本，并解决了一些加载config时候的报错，
`"""Download LeWM checkpoints from Hugging Face and convert to eval-compatible format."""

import argparse, json, os
import torch
import stable_pretraining as spt
from huggingface_hub import hf_hub_download
from jepa import JEPA
from module import ARPredictor, Embedder, MLP

HF_REPOS = {
    "pusht":    {"repo": "quentinll/lewm-pusht",    "output_dir": "pusht"},
    "tworooms": {"repo": "quentinll/lewm-tworooms", "output_dir": "two-room"},
    "cube":     {"repo": "quentinll/lewm-cube",     "output_dir": "cube"},
    "reacher":  {"repo": "quentinll/lewm-reacher",  "output_dir": "reacher"},
}

def build_model_from_config(cfg):
    enc = cfg["encoder"]
    encoder = spt.backbone.utils.vit_hf(
        enc["size"], patch_size=enc["patch_size"],
        image_size=enc["image_size"], pretrained=False, use_mask_token=False,
    )
    predictor_cfg = {k: v for k, v in cfg["predictor"].items() if k != "_target_"}
    action_encoder_cfg = {k: v for k, v in cfg["action_encoder"].items() if k != "_target_"}
    projector_cfg = {
        k: v for k, v in cfg["projector"].items()
        if k not in ["_target_", "norm_fn"]
    }
    pred_proj_cfg = {
        k: v for k, v in cfg["pred_proj"].items()
        if k not in ["_target_", "norm_fn"]
    }
    return JEPA(
        encoder=encoder,
        predictor=ARPredictor(**predictor_cfg),
        action_encoder=Embedder(**action_encoder_cfg),
        projector=MLP(**projector_cfg, norm_fn=torch.nn.BatchNorm1d),
        pred_proj=MLP(**pred_proj_cfg, norm_fn=torch.nn.BatchNorm1d),
    )

def convert_task(task, cache_dir):
    # info = HF_REPOS[task]
    # config_path = hf_hub_download(repo_id=info["repo"], filename="config.json")
    # weights_path = hf_hub_download(repo_id=info["repo"], filename="weights.pt")
    config_path = "/data/cube/config.json"
    weights_path = "/data/cube/weights.pt"
    with open(config_path) as f:
        cfg = json.load(f)
    model = build_model_from_config(cfg)
    model.load_state_dict(torch.load(weights_path, map_location="cpu", weights_only=True))
    model.eval()
    # out_dir = os.path.join(cache_dir, info["output_dir"])
    out_dir = "/data/cube_ckpt"
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "lewm_object.ckpt")
    torch.save(model, out_path)
    print(f"Saved: {out_path}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--tasks", nargs="+", choices=list(HF_REPOS), default=list(HF_REPOS))
    parser.add_argument("--cache-dir", default=os.environ.get("STABLEWM_HOME", os.path.expanduser("~/.stable-wm")))
    args = parser.parse_args()
    for task in args.tasks:
        convert_task(task, args.cache_dir)`
